Make ShardedTensor ctor more inline with torch.Tensor ctor

pritamdamania87 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #72164

torch.Tensor ctor creates an empty tensor and this PR makes
ShardedTensor on par with that.

In particular we remove TensorInitParams and instead always a create an empty
tensor and then fill it in for things like ones, zeros, full etc. This is
inline with torch.ones etc. as well since even for those APIs we first create
an empty tensor and then fill it out.

Differential Revision: [D33934603](https://our.internmc.facebook.com/intern/diff/D33934603/)

pritamdamania committed:
Make ShardedTensor ctor more inline with torch.Tensor ctor

torch.Tensor ctor creates an empty tensor and this PR makes
ShardedTensor on par with that.

In particular we remove TensorInitParams and instead always a create an empty
tensor and then fill it in for things like ones, zeros, full etc. This is
inline with torch.ones etc. as well since even for those APIs we first create
an empty tensor and then fill it out.

Differential Revision: [D33934603](https://our.internmc.facebook.com/intern/diff/D33934603/)

[ghstack-poisoned]

albanD commented on torch/nn/init.py:
```diff
@@ -190,6 +190,8 @@ def constant_(tensor: Tensor, val: float) -> Tensor:
         >>> w = torch.empty(3, 5)
         >>> nn.init.constant_(w, 0.3)
     """
+    if has_torch_function_variadic(tensor):
```
Isn't this already catching the `.fill_()` function? Why do we want this here?

pritamdamania87 replied:
I think this calls tensor.fill_() under the hood and not torch.fill_(). I tried a simple test and it doesn't seem to work:

```
import torch

class MyTensor():

    def __torch_function__(self, func, types, args=(), kwargs=None):
        print(f'called: {func}')

t = MyTensor()
torch.nn.init.constant_(t, 0)
```

The error is:
```
'MyTensor' object has no attribute 'fill_'
---------------------------------------------------------------------------
AttributeError                            Traceback (most recent call last)
 in <module>
      7 
      8 t = MyTensor()
----> 9 torch.nn.init.constant_(t, 0)
torch/nn/init.py in constant_(tensor, val)
    191         >>> nn.init.constant_(w, 0.3)
    192     """
--> 193     return _no_grad_fill_(tensor, val)
    194 
    195 
torch/nn/init.py in _no_grad_fill_(tensor, val)
     60 def _no_grad_fill_(tensor, val):
     61     with torch.no_grad():
---> 62         return tensor.fill_(val)
     63 
     64 
AttributeError: 'MyTensor' object has no attribute 'fill_'
```

albanD replied:
Right if you're not using a Tensor subclass, you don;t get any method "for free" you have to add them all. That is expected.

wanchaol commented on torch/distributed/_sharded_tensor/ops/init.py:
```diff
@@ -86,3 +86,42 @@ def kaiming_uniform_(types, args=(), kwargs=None, pg=None):
     return sharded_tensor
+
+@sharded_op_impl(torch.nn.init.normal_)
+def normal_(types, args=(), kwargs=None, pg=None):
```
it seems that we already got `normal_` in line 33?

wanchaol reviewed (approved):
looks good! One nit is that there seems to be a redefinition with `normal_`

pritamdamania committed:
Update on "Make ShardedTensor ctor more inline with torch.Tensor ctor"


torch.Tensor ctor creates an empty tensor and this PR makes
ShardedTensor on par with that.

In particular we remove TensorInitParams and instead always a create an empty
tensor and then fill it in for things like ones, zeros, full etc. This is
inline with torch.ones etc. as well since even for those APIs we first create
an empty tensor and then fill it out.

Differential Revision: [D33934603](https://our.internmc.facebook.com/intern/diff/D33934603/)

[ghstack-poisoned]

pritamdamania committed:
Update on "Make ShardedTensor ctor more inline with torch.Tensor ctor"


torch.Tensor ctor creates an empty tensor and this PR makes
ShardedTensor on par with that.

In particular we remove TensorInitParams and instead always a create an empty
tensor and then fill it in for things like ones, zeros, full etc. This is
inline with torch.ones etc. as well since even for those APIs we first create
an empty tensor and then fill it out.

Differential Revision: [D33934603](https://our.internmc.facebook.com/intern/diff/D33934603/)

[ghstack-poisoned]
